src/number_theory/fermat_psp.lean

```diff
@@ -0,0 +1,548 @@
+  nat.coprime n b :=
+begin
+  by_cases h₃ : 2 ≤ n,
+
+  { -- To prove that `n` is coprime with `b`, we we need to show that for all prime factors of `n`,
+    -- we can derive a contradiction if `n` divides `b`.
+    apply nat.coprime_of_dvd,
+
+    -- If `k` is a prime number that divides both `n` and `b`, then we know that `n = m * k` and
+    -- `b = j * k` for some natural numbers `m` and `j`. We substitute these into the hypothesis.
+    rintros k hk ⟨m, rfl⟩ ⟨j, rfl⟩,
+
+    -- Because prime numbers do not divide 1, it suffices to show that `k ∣ 1` to prove a
+    -- contradiction
+    apply nat.prime.not_dvd_one hk,
+
+    -- Since `n` divides `b ^ (n - 1) - 1`, `k` also divides `b ^ (n - 1) - 1`
+    replace h := dvd_of_mul_right_dvd h,
+
+    -- Because `k` divides `b ^ (n - 1) - 1`, if we can show that `k` also divides `b ^ (n - 1)`,
+    -- then we know `k` divides 1.
+    rw [nat.dvd_add_iff_right h, nat.sub_add_cancel (nat.one_le_pow _ _ h₂)],
+
+    -- Since `k` divides `b`, `k` also divides any power of `b` except `b ^ 0`. Therefore, it
+    -- suffices to show that `n - 1` isn't zero. However, we know that `n - 1` isn't zero because we
+    -- assumed `2 ≤ n` when doing `by_cases`.
+    refine dvd_of_mul_right_dvd (dvd_pow_self (k * j) _),
+    linarith },
+
+  -- If `n = 1`, then it follows trivially that `n` is coprime with `b`.
+  { rw (show n = 1, by linarith),
+    norm_num }
+end
+
+lemma probable_prime_iff_modeq (n : ℕ) {b : ℕ} (h : 1 ≤ b) :
+  probable_prime n b ↔ b^(n - 1) ≡ 1 [MOD n] :=
+begin
+  have : 1 ≤ b ^ (n - 1) := one_le_pow_of_one_le h (n - 1), -- For exact_mod_cast
+  rw nat.modeq.comm,
+  split,
+  { intro h₁,
+    apply nat.modeq_of_dvd,
+    exact_mod_cast h₁, },
+  { intro h₁,
+    exact_mod_cast nat.modeq.dvd h₁, },
+end
+
+/--
+If `n` is a Fermat pseudoprime to base `b`, then `n` is coprime with `b`, assuming that `b` is
+positive.
+
+This lemma is a small wrapper based on `coprime_of_probable_prime`
+-/
+lemma coprime_of_fermat_psp (n b : ℕ) (h : fermat_psp n b) (h₁ : 1 ≤ b) : nat.coprime n b :=
+begin
+  rcases h with ⟨hp, hn₁, hn₂⟩,
+  exact coprime_of_probable_prime n b hp (by linarith) h₁,
+end
+
+/--
+All composite numbers are Fermat pseudoprimes to base 1.
+-/
+lemma pseudoprime_of_base_one (n : ℕ) (h₁ : 1 < n) (h₂ : ¬nat.prime n) : fermat_psp n 1 :=
+begin
+  split,
+  { have h : 0 = 1^(n - 1) - 1 := by norm_num,
+    show n ∣ 1^(n - 1) - 1, from h ▸ (dvd_zero n) },
+  { exact ⟨h₂, h₁⟩ }
+end
+
+-- For lemmas that are needed to prove statements in this file, but aren't directly related to
+-- Fermat pseudoprimes
+section helper_lemmas
+
+private lemma odd_of_prime_gt_two (p : ℕ) (h : nat.prime p) (hp : 2 < p) : odd p :=
+begin
+  rw [nat.odd_iff_not_even, even_iff_two_dvd],
+  rw nat.prime_dvd_prime_iff_eq nat.prime_two h,
+  exact ne_of_lt hp
+end
+
+private lemma odd_pow_lem (a : ℤ) (n k : ℕ) (h₁ : k ∣ n) (h₂ : odd (n / k)) : a^k + 1 ∣ a^n + 1 :=
```

Comment:
For some reason that I don't remember anymore, I decided to let `a` be an integer rather than a natural number. It seems to still work if I use `exact_mod_cast` though.

Reply:
`odd_pow_lem` has been replaced with a lemma added in #17996